### PR TITLE
Feat: auto update

### PR DESCRIPTION
**Pull request by DJMcNab**
This adds a command to automatically update the server process.

I just want to check that this is something which is wanted, and there are a couple of issues - hence the draft PR:
 - The output channel isn't working properly, for some reason
 - Update on startup isn't setup yet.

### CODE REVIEW

File: editors/code/src/commands/update.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+async function run(repo: string) {
+    try {
+        // Handle windows not allowing
+        await renameAsync(join(homedir(), '.cargo/bin/ra_lsp_server'), newPath);
+    } catch (error) {
+        // Fall through to next
+    }
+    window.showInformationMessage('Installing newer version');
+    const spawned = spawn(
+        `cargo install --git ${repo} ra_lsp_server --force ${Server.config.serverUpdateOptions.arguments}`
```
Comment (bjorn3) [inline editors/code/src/commands/update.ts]: This wont update the vscode extension itself, which may be needed for new features. It also rebuilds every single dependency of ra_lsp_server every time you try to update it, making very slow.

**Comment by DJMcNab** [comment]
Yeah, those were both limitations I was aware of.

I'm not sure what a good way to allow incremental compilation is - I suppose we would have to do a shallow clone.

Updating the vscode extension automatically is infeasible - the only real approach would be to start publishing it, but if we did that we'd probably want to have prebuilt binaries.

Sorry about the full comment - my app doesn't want to let me reply to the inline comment

**Comment by lnicola** [comment]
It's probably a better idea to publish the extension to Marketplace and bundle language server binaries with it.

**Comment by DJMcNab** [comment]
The issue with that is that we would need to bundle a Linux, Macos and windows binary with all the versions. Ideally, we would build and deploy the binaries to S3 or something, and the extension would download the latest (basically what rls does atm, but rls has rustup to help)

**Comment by DJMcNab** [comment]
Of course the alternative is getting it running in wasm, but that would require some more deep changes (including probably moving to lazy file loading) - then building a simple pass-through node script

**Comment by matklad** [comment]
I am on the fence about it. I think we are not quite ready for auto-updates yet: auto update should update to a known good version, but we don't have any release trains to figure out what the good version is. Are we ready to setup proper release process at this moment? I am not sure: the main value of rust-analyzer is in experimentation and determining what the right architecture is, and not in the direct usability improvements. 

I think that having a command to (manually) update would be useful, but I am not sure that `cargo install` from the git repo is the way to go here, as it doesn't do caching. Could we perhaps clone the repo to https://github.com/microsoft/vscode/blob/cb70fd5ef0d0201a64dff3b878a5574dfe2bb1e9/src/vs/vscode.d.ts#L4846 ? This seems pretty complex though.

Overall, it seems like the complexity here is not worth the benefit yet. On the other hand, it might be interesting to publish this as a separate "update rust-analyzer" extension.

**Comment by matklad** [comment]
So, let's close this for now!

**Comment by DJMcNab** [comment]
Agreed.

**Comment by gilescope** [comment]
Now we have pre-built binaries... is it time to revisit this @DJMcNab, @matklad?

(As the changelog is published every week, that drives demand for people to upgrade every week, so it would be great if that was as seamless as possible).

**Comment by matklad** [comment]
https://github.com/rust-analyzer/rust-analyzer/issues/2988

**Pull request closed without merging**